api_keys/file.go

```diff
@@ -0,0 +1,84 @@
+package apikeys
+
+import (
+	"encoding/csv"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+)
+
+type APIKeysFromFile struct {
+	FileName    string
+	users       map[string]string
+	lastUpdated time.Time
+	mu          sync.Mutex
+}
+
+func (k *APIKeysFromFile) readCSVToMap() (map[string]string, error) {
+	// Open the CSV file
+	file, err := os.Open(k.FileName)
+	if err != nil {
+		return nil, fmt.Errorf("error opening file: %v", err)
+	}
+	defer file.Close()
+
+	// Create a CSV reader
+	reader := csv.NewReader(file)
+
+	// Read all the records from the CSV
+	records, err := reader.ReadAll()
+	if err != nil {
+		return nil, fmt.Errorf("error reading CSV: %v", err)
+	}
+
+	apiMap := make(map[string]string)
+
+	// Skip the header and populate the map
+	for _, record := range records {
+		if len(record) < 2 {
+			continue // Skip records with insufficient fields
+		}
+		apiKey := record[0]
+		user := record[1]
+		apiMap[apiKey] = user
+	}
+
+	return apiMap, nil
+}
+
+func (k *APIKeysFromFile) GetDetailsByKey(key string) (APIKeyDetails, bool) {
+	k.mu.Lock()
+	defer k.mu.Unlock()
+
+	if k.users == nil || time.Since(k.lastUpdated) > 30*time.Second {
+		users, ok := k.readCSVToMap()
```

Comment:
If `readCSVToMap()` is called at struct initialisation, the exclusion lock won't be necessary unless for writes.